### PR TITLE
refactor(server): extract session-resolution + capability-gate helpers (#4773)

### PR DESCRIPTION
**Pull request by blamechris**
## Summary

Collapses the resolve-then-error and capability-gate boilerplate that the DRY/SOLID sweep flagged in #4773. Three new helpers in `handler-utils.js`:

- **`sendSessionError(ws, ctx, message)`** — one-liner for the canonical `session_error` envelope. Centralises the wire shape so a future schema tweak (adding `code`, `recoverable`, `sessionId`, etc.) lands in one place.
- **`resolveSessionOrError(ws, ctx, msg, client)`** — wraps `resolveSession`, emits the canonical \"No active session\" envelope on miss, returns `null`. Callers stay `if (!entry) return`.
- **`requireSessionMethod(ws, ctx, entry, method, message)`** — wraps the `typeof entry.session.setX !== 'function'` capability gate; emits the provider-not-supported envelope and returns `false` on miss.

Migrated call sites (actual landed counts — Copilot caught the mismatch in my first description):
- **8** sites adopted `resolveSessionOrError` (7 in `settings-handlers.js`: `handleSkillActivate`, `handleSkillDeactivate`, `handleSkillTrustAccept`, `handleSkillTrustGrant`, `handleSetThinkingLevel`, `handleSetPromptEvaluatorSkipPattern`, `handleSetPermissionRules`; 1 in `per-session-settings.js`).
- **6** sites adopted `requireSessionMethod` (5 in `settings-handlers.js`: `handleSkillActivate`, `handleSkillDeactivate`, `handleSetThinkingLevel`, `handleSetPromptEvaluatorSkipPattern`, `handleSetPermissionRules`; 1 in `per-session-settings.js`).
- **~24** inline `session_error` envelopes collapsed to `sendSessionError(ws, ctx, …)` (includes the per-rule validation messages in `handleSetPermissionRules`, both checkpoint-handlers entry points that resolve sessions via `client.activeSessionId` directly rather than `resolveSession`, and the plan-mode capability gate that lives inside an `if (entry)` block).

The 13/6 counts in #4773 were based on the issue's inventory pattern; the 7 + 1 = 8 actual `resolveSessionOrError` migrations are smaller because (a) the two checkpoint-handlers sites use `ctx.sessionManager.getSession(sid)` directly (not `resolveSession`) and so migrated to `sendSessionError` instead, and (b) three settings-handlers sites use the `if (entry) {…}` pattern and don't bail on a null entry, so they retain `resolveSession`. Behaviour-preserving — see below.

Net diff: **+251 LOC** (+193 LOC of tests + helper JSDoc, −92 LOC of inline boilerplate removed).

What **didn't** change:
- The 3 sites that use `resolveSession` directly with the `if (entry)…` pattern (handleSetModel, handleSetPermissionMode top-level, handleListSkills) — they tolerate `null` entry without bailing, so they wouldn't benefit from `resolveSessionOrError`.
- Provider-specific error codes routed through `sendError(ws, requestId, code, message)` — those carry structured codes and stay on the existing helper.
- Behaviour: every existing handler test passes unmodified.

## Behavior preservation

The helpers route through `ctx.send` (not `ws.send` directly), preserving compatibility with the existing handler test mocks that monkey-patch `ctx.send` and inspect `ctx._sent`. In production `ctx.send` ultimately calls `ws.send(JSON.stringify(msg))` via `WsServer._send`, so the wire output is byte-identical to the inline pattern it replaces.

Verification:
- All existing handler tests pass unchanged (settings-handlers, checkpoint-handlers, per-session-settings, settings-handlers-multi-client-broadcast, settings-handlers-community-scan-order, plus the ws-server provider capability gate tests).
- 16 new boundary tests in `tests/handler-utils.test.js` cover the three helpers (hit/miss paths, null/undefined safety, binding violations, method-exists-but-not-function edge case).
- Full server suite passes the same set of tests as baseline; the 6 known failures (service/tunnel binary missing, WebTaskManager flake) are pre-existing and unrelated.

## Test plan

- [x] \`cd packages/server && npm test\` — same baseline pass/fail as before the refactor
- [x] Boundary tests in tests/handler-utils.test.js cover sendSessionError, resolveSessionOrError, requireSessionMethod
- [x] Migrated handler tests (handlers/settings-handlers.test.js, handlers/checkpoint-handlers.test.js, per-session-settings.test.js) all pass without modification — proves the helpers are behaviour-preserving wrappers

Closes #4773

### CODE REVIEW

File: packages/server/src/handler-utils.js [hunk 1]
```diff
@@ -398,6 +398,94 @@ export function resolveSession(ctx, msg, client) {
...
+ * @returns {object|null} Session entry on hit, null after emitting on miss
+ */
+export function resolveSessionOrError(ws, ctx, msg, client) {
+  const entry = resolveSession(ctx, msg, client)
+  if (!entry) {
+    sendSessionError(ws, ctx, 'No active session')
+    return null
+  }
+  return entry
+}
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/src/handler-utils.js:456]: **FIX (description-only)** — you're right, the counts in the PR description didn't match what actually landed. Just updated the description with the accurate breakdown: 8 `resolveSessionOrError` migrations (7 settings + 1 per-session-settings), 6 `requireSessionMethod`, ~24 inline `sendSessionError`. The two checkpoint-handlers sites in the issue inventory call `ctx.sessionManager.getSession(sid)` directly instead of `resolveSession` (they use `client.activeSessionId` not `msg.sessionId || client.activeSessionId`), so they migrated to `sendSessionError` rather than `resolveSessionOrError` — pre-refactor pattern was incompatible. The three remaining settings-handlers sites use the `if (entry) {…}` pattern that tolerates a null entry, so migrating them would change behaviour. No code change needed.
